File: test/smoke/run_presenter.rb
# frozen_string_literal: true

# Host smoke test for the presenter view (PresenterDeck) — the Ruby port of
# test/smoke/run-presenter.mjs. Boots the vendored lilac.wasm on wasmtime-rb
# (dommy-js-wasmtime, QuickJS + Dommy) and checks the presenter view renders the
# current slide, the upcoming previews, and the speaker notes, and that
# navigating advances all of them.
#
# Cross-window sync (BroadcastChannel) is NOT exercised: it's removed so
# PresenterDeck#setup_sync no-ops (a browser-only two-window feature). This smoke
# covers the render + local navigation contract.
#
# Run: ruby test/smoke/run_presenter.rb   (or `make smoke`)

require "json"

require "dommy"
require "dommy/js/wasmtime" # provided via Gemfile (gem "dommy-js-wasmtime")

ROOT = File.expand_path("../..", __dir__)
WASM = File.join(ROOT, "vendor/lilac/lilac.wasm")
SOURCES = %w[bus slides renderer player vtt_track slide deck].map { |n| File.join(ROOT, "lib/#{n}.rb") }

# 5 slides; slide 0 carries notes, slide 1 is a poem (player) slide.
SLIDES_JSON = JSON.generate(
  "metadata" => { "title" => "Smoke", "language" => "ja", "total_slides" => 5 },
  "slides" => [
    { "index" => 0, "title" => "Intro", "html" => "<h1>Intro</h1><p>x</p>", "layout" => "two-column",
      "regions" => ["<h1>Intro</h1>", "<p>x</p>"], "speaker_notes" => "発表者メモ0", "dynamic" => false, "player" => nil },
    { "index" => 1, "title" => "A", "html" => "<h1>A</h1>", "layout" => nil, "speaker_notes" => nil, "dynamic" => false,
      "player" => { "audio" => "media/a.mp3", "vtt" => "WEBVTT\n\n00:00.000 --> 00:01.000\nあ\n" } },
    { "index" => 2, "title" => "B", "html" => "<h1>B</h1>", "layout" => nil, "speaker_notes" => nil,
      "dynamic" => false, "player" => nil },
    { "index" => 3, "title" => "C", "html" => "<h1>C</h1>", "layout" => nil, "speaker_notes" => nil,
      "dynamic" => false, "player" => nil },
    { "index" => 4, "title" => "D", "html" => "<h1>D</h1>", "layout" => nil, "speaker_notes" => nil,
      "dynamic" => false, "player" => nil }
  ]
)

SETUP_JS = <<~JS
  delete globalThis.BroadcastChannel;
  globalThis.fetch = async (url) =>
    new Response(#{SLIDES_JSON.to_json}, { status: 200, headers: { "Content-Type": "application/json" } });
JS

FAILS = []
def check(cond, msg)
  return if cond

  FAILS << msg
  warn "FAIL: #{msg}"
end

vm = Dommy::Js::Wasmtime.boot(wasm: WASM, html: File.read(File.join(ROOT, "viewer/presenter.html")),
                              sources: SOURCES, entrypoint: "Lilac.start") do |engine|
  engine.eval(SETUP_JS)
end
DOC = vm.document
$engine = vm.engine
$vm = vm

def q(sel) = DOC.query_selector(sel)
def qa(sel) = DOC.query_selector_all(sel)

def dispatch(js)
  $engine.eval(js)
  $vm.drain_async!
end

# Initial: current = slide 0, 3 upcoming previews (2/3/4), notes of slide 0.
check(q(".preview-current").inner_html.include?("Intro"), "current preview is slide 0")
check(q(".preview-current .slide-body").get_attribute("data-layout") == "two-column",
      "current preview applies the slide layout (two-column)")
check(qa(".preview-current .slide-region").length == 2,
      "current preview renders regions (got #{qa('.preview-current .slide-region').length})")
check(q(".counter").text_content == "1 / 5", "counter '1 / 5' (got #{q('.counter').text_content})")
previews = qa(".preview-list .preview-small")
check(previews.length == 3, "3 upcoming previews (got #{previews.length})")
check(qa(".preview-num")[0].text_content == "2",
      "first upcoming numbered 2 (got #{qa('.preview-num')[0]&.text_content})")
check(previews[0].inner_html.include?("A"), "first upcoming is slide 1 (A)")
check(q(".notes-body").text_content.include?("発表者メモ0"), "notes of current slide shown")
check(q(".presenter-player").class_list.contains?("lil-hidden"), "player remote hidden on non-poem slide")
puts "[ok] presenter renders current + upcoming + notes"

# Advance with ArrowRight: everything tracks @index.
dispatch('document.dispatchEvent(new KeyboardEvent("keydown", { key: "ArrowRight", bubbles: true }))')
check(q(".preview-current").inner_html.include?("A"), "current advanced to slide 1 (A)")
check(q(".counter").text_content == "2 / 5", "counter '2 / 5' (got #{q('.counter').text_content})")
check(qa(".preview-num")[0].text_content == "3",
      "upcoming now starts at 3 (got #{qa('.preview-num')[0]&.text_content})")
check(q(".notes-body").text_content.include?("ノートはありません"), "no-notes placeholder for slide 1")
puts "[ok] navigation advances current + upcoming + notes"

# Slide 1 is a poem slide → the remote-control appears with 3 buttons.
check(!q(".presenter-player").class_list.contains?("lil-hidden"), "player remote shown on poem slide")
check(qa(".presenter-player .controls button").length == 3, "remote has play/pause/reset")
dispatch('document.querySelector(\'.presenter-player [data-on-click="play"]\').dispatchEvent(new MouseEvent("click", { bubbles: true }))')
dispatch('document.querySelector(\'.presenter-player [data-on-click="reset"]\').dispatchEvent(new MouseEvent("click", { bubbles: true }))')
puts "[ok] poem slide shows the play/pause/reset remote + indicator"

# Near the end: fewer upcoming previews (clamped, no wrap).
dispatch('document.querySelector(\'[data-on-click="go_next"]\').dispatchEvent(new MouseEvent("click", { bubbles: true }))') # -> 2 (B)
dispatch('document.querySelector(\'[data-on-click="go_next"]\').dispatchEvent(new MouseEvent("click", { bubbles: true }))') # -> 3 (C)
check(q(".counter").text_content == "4 / 5", "counter '4 / 5' (got #{q('.counter').text_content})")
check(qa(".preview-list .preview-small").length == 1,
      "only 1 upcoming near end (got #{qa('.preview-list .preview-small').length})")
puts "[ok] upcoming clamps at the end (no wrap)"

if FAILS.empty?
  puts "ALL OK"
else
  warn "\n#{FAILS.length} check(s) failed"
  exit 1
end

File: test/smoke/run-presenter.mjs
// Host smoke test for the presenter view (PresenterDeck). Boots the vendored
// lilac.wasm under happy-dom and checks that the presenter view renders the
// current slide, the upcoming previews, and the speaker notes, and that
// navigating advances all of them.
//
// Cross-window sync (BroadcastChannel) is NOT exercised here: happy-dom has no
// BroadcastChannel, so PresenterDeck#setup_sync no-ops (verified indirectly by
// boot succeeding). Sync is a browser-only, two-window feature — checked
// manually. This smoke covers the render + local navigation contract.
//
// Run: npm run smoke:presenter  (node --experimental-wasm-exnref)

import { readFile } from "node:fs/promises";
import { fileURLToPath } from "node:url";

// See run-node.mjs: remove Node's BroadcastChannel so the deck runs the
// no-sync path (cross-window sync is browser-only and would keep the event
// loop alive here).
delete globalThis.BroadcastChannel;

const { Window } = await import("happy-dom");
const dom = new Window({ url: "https://test.local/" });
globalThis.document = dom.document;
globalThis.window = dom;
globalThis.requestAnimationFrame = dom.requestAnimationFrame.bind(dom);
globalThis.cancelAnimationFrame = dom.cancelAnimationFrame.bind(dom);
globalThis.MutationObserver = dom.MutationObserver;
globalThis.URL = dom.URL ?? globalThis.URL;
globalThis.Blob = dom.Blob ?? globalThis.Blob;

// 5 slides; slide 0 carries notes. No player slides needed for this view.
const SLIDES_JSON = JSON.stringify({
  metadata: { title: "Smoke", language: "ja", total_slides: 5 },
  slides: [
    { index: 0, title: "Intro", html: "<h1>Intro</h1><p>x</p>", layout: "two-column",
      regions: ["<h1>Intro</h1>", "<p>x</p>"],
      speaker_notes: "発表者メモ0", dynamic: false, player: null },
    // Slide 1 is a poem (player) slide → presenter shows the remote-control.
    { index: 1, title: "A", html: "<h1>A</h1>", layout: null,
      speaker_notes: null, dynamic: false,
      player: { audio: "media/a.mp3", vtt: "WEBVTT\n\n00:00.000 --> 00:01.000\nあ\n" } },
    { index: 2, title: "B", html: "<h1>B</h1>", layout: null,
      speaker_notes: null, dynamic: false, player: null },
    { index: 3, title: "C", html: "<h1>C</h1>", layout: null,
      speaker_notes: null, dynamic: false, player: null },
    { index: 4, title: "D", html: "<h1>D</h1>", layout: null,
      speaker_notes: null, dynamic: false, player: null },
  ],
});

globalThis.fetch = async (url) => {
  const s = String(url);
  if (s.endsWith("slides.json")) {
    return new Response(SLIDES_JSON, { headers: { "Content-Type": "application/json" } });
  }
  const path = fileURLToPath(new URL(s, import.meta.url));
  return new Response(await readFile(path), {
    headers: { "Content-Type": s.endsWith(".wasm") ? "application/wasm" : "text/plain" },
  });
};

// Mirror viewer/presenter.html body.
document.body.innerHTML = `
  <main class="presenter" data-component="PresenterDeck">
    <header class="presenter-bar">
      <button type="button" data-on-click="go_prev">←</button>
      <span class="counter" data-text="@counter"></span>
      <button type="button" data-on-click="go_next">→</button>
    </header>
    <div class="presenter-grid">
      <section class="presenter-current">
        <div class="preview preview-current">
          <div class="slide-body" data-attr-data-layout="@current_layout">
            <div class="slide-regions" data-each="@current_regions" data-key="n">
              <div class="slide-region slide-html" data-unsafe-html="html"></div>
            </div>
          </div>
        </div>
        <div class="presenter-player" data-show="@has_player">
          <div class="progress" data-ref="progress">
            <div class="progress-played" data-ref="progress_played"></div>
          </div>
          <div class="controls">
            <button type="button" data-on-click="play">再生する</button>
            <button type="button" data-on-click="pause">止める</button>
            <button type="button" data-on-click="reset">最初に戻る</button>
          </div>
        </div>
      </section>
      <section class="presenter-next">
        <div class="preview-list" data-each="@upcoming" data-key="n">
          <div class="preview preview-small">
            <span class="preview-num" data-text="num"></span>
            <div class="preview-body slide-html" data-unsafe-html="html"></div>
          </div>
        </div>
      </section>
    </div>
    <section class="presenter-notes">
      <div class="notes-body" data-text="@notes"></div>
    </section>
  </main>
  <p id="boot-error" class="error" hidden></p>
`;

const { bootRuby } = await import("../../src/ruby_runtime.js");
await bootRuby();
const tick = (ms = 30) => new Promise((r) => setTimeout(r, ms));
await tick();

const q = (sel) => document.querySelector(sel);
const qa = (sel) => Array.from(document.querySelectorAll(sel));
function check(cond, msg) {
  if (!cond) { console.error("FAIL:", msg); process.exit(1); }
}

// Initial: current = slide 0, 3 upcoming previews (2/3/4), notes of slide 0.
check(q(".preview-current").innerHTML.includes("Intro"), "current preview is slide 0");
// presenter-current uses the same slide-body/regions layout as the normal view.
check(q(".preview-current .slide-body").getAttribute("data-layout") === "two-column",
      "current preview applies the slide layout (two-column)");
check(qa(".preview-current .slide-region").length === 2,
      `current preview renders regions (got ${qa(".preview-current .slide-region").length})`);
check(q(".counter").textContent === "1 / 5", `counter "1 / 5" (got ${q(".counter").textContent})`);
const previews = qa(".preview-list .preview-small");
check(previews.length === 3, `3 upcoming previews (got ${previews.length})`);
check(qa(".preview-num")[0].textContent === "2", `first upcoming numbered 2 (got ${qa(".preview-num")[0].textContent})`);
check(previews[0].innerHTML.includes("A"), "first upcoming is slide 1 (A)");
check(q(".notes-body").textContent.includes("発表者メモ0"), "notes of current slide shown");
// Slide 0 is not a poem slide → the remote-control is hidden.
check(q(".presenter-player").classList.contains("lil-hidden"), "player remote hidden on non-poem slide");
console.log("[ok] presenter renders current + upcoming + notes");

// Advance with ArrowRight: everything tracks @index.
document.dispatchEvent(new dom.window.KeyboardEvent("keydown", { key: "ArrowRight", bubbles: true }));
await tick();
check(q(".preview-current").innerHTML.includes("A"), "current advanced to slide 1 (A)");
check(q(".counter").textContent === "2 / 5", `counter "2 / 5" (got ${q(".counter").textContent})`);
check(qa(".preview-num")[0].textContent === "3", `upcoming now starts at 3 (got ${qa(".preview-num")[0].textContent})`);
check(q(".notes-body").textContent.includes("ノートはありません"), "no-notes placeholder for slide 1");
console.log("[ok] navigation advances current + upcoming + notes");

// Slide 1 is a poem slide → the remote-control appears with 3 buttons, and
// the buttons are clickable (no-op here: BroadcastChannel removed above).
check(!q(".presenter-player").classList.contains("lil-hidden"), "player remote shown on poem slide");
check(qa(".presenter-player .controls button").length === 3, "remote has play/pause/reset");
q('.presenter-player [data-on-click="play"]').dispatchEvent(new dom.window.MouseEvent("click", { bubbles: true }));
q('.presenter-player [data-on-click="reset"]').dispatchEvent(new dom.window.MouseEvent("click", { bubbles: true }));
await tick();
console.log("[ok] poem slide shows the play/pause/reset remote + indicator");

// Near the end: fewer upcoming previews (clamped, no wrap).
q('[data-on-click="go_next"]').dispatchEvent(new dom.window.MouseEvent("click", { bubbles: true })); // -> 2 (B)
await tick();
q('[data-on-click="go_next"]').dispatchEvent(new dom.window.MouseEvent("click", { bubbles: true })); // -> 3 (C)
await tick();
check(q(".counter").textContent === "4 / 5", `counter "4 / 5" (got ${q(".counter").textContent})`);
check(qa(".preview-list .preview-small").length === 1, `only 1 upcoming near end (got ${qa(".preview-list .preview-small").length})`);
console.log("[ok] upcoming clamps at the end (no wrap)");

console.log("ALL OK");
